Added NPC and Object search functions

StickToTheScript opened the pull request:
Added more NPC and Object searching functions and a function to grab the main screen bounds.

StickToTheScript committed:
Added NPC and Object searching functions

- Added NPC functions for getting NPC tiles and MS coordinates.
- Added Object functions for getting all visible objects of a specific ID or name and returning their tiles.

StickToTheScript committed:
Added Mainscreen.simba

StickToTheScript committed:
Allowed for other object types in RSObject.GetGameObjectOnScreen

StickToTheScript committed:
Added function to return ground items ms coords

StickToTheScript committed:
Added R_NPCGetMSCoordsClosest

Wanted a function to return the closest NPC's Mainscreen coordinates.

StickToTheScript committed:
Update Objects.simba

Added functions to return screen coordinates.

StickToTheScript committed:
Update NPC.simba

StickToTheScript committed:
Updated Ground.simba to use Mainscreen bounds

franher007 commented on Reflection/Ground.simba:
```diff
@@ -10,21 +10,52 @@ Example:
-  if Length(Items) < 1 then
+  MSBounds := R_MainscreenBounds;
+  if High(Items) < 1 then
     Exit;
```
You could invert the order of these calls to sometimes save a call

StickToTheScript replied:
Fixed! Thanks for that find!

franher007 commented on Reflection/Ground.simba:
```diff
@@ -10,21 +10,52 @@ Example:
-    GetClientDimensions(W, H);
     ResultTile := RSTile(Items[i]).ToLocal.TileToMS;
-    if (not ((ResultTile.X < 0) or (ResultTile.X > W) or (ResultTile.Y < 0) or (ResultTile.Y > H))) then
+    if (not ((ResultTile.X > MSBounds.X1) and (ResultTile.X < MSBounds.X2) and (ResultTile.Y > MSBounds.Y1) and (ResultTile.Y < MSBounds.Y2))) then
```
Something i use in my scripts:

```
function TPoint.InBox(Box: TBox): Boolean; constref;
begin
  Result := PointInBox(Self, Box);
end; 
```
Then you can just ask ResultTile.InBox(MSBounds)

StickToTheScript replied:
I'm surprised that I forgot about that function.. Thanks!

franher007 commented on Reflection/Ground.simba:
```diff
@@ -10,21 +10,52 @@ Example:
+  for I := 0 to High(Items) do
+  begin
+    ResultTile := RSTile(Items[i]).ToLocal.TileToMS;
+    if (not ((ResultTile.X > MSBounds.X1) and (ResultTile.X < MSBounds.X2) and (ResultTile.Y > MSBounds.Y1) and (ResultTile.Y < MSBounds.Y2))) then
```
Same suggestion as above :)

franher007 commented on Reflection/NPC.simba:
```diff
@@ -201,3 +222,234 @@ begin
+
+   writeln(R_NPCGetMSCoordsSort('Chicken'));
+*)
+Function R_NPCGetMSCoordsSort(Name: String): TPointArray;
```
I'm thinking here you could do something like:

```
Function R_NPCGetMSCoordsSort(Name: String): TPointArray;
var
  Tiles : TPointArray;
begin
  Tiles := R_NPCGetMSCoords(Name);
  Tiles.sort(R_MainscreenBounds.Middle);
  Result := Tiles;
end;
```

JGBMichalski replied:
Haha. I implemented that but must've accidentally written over my changes.

StickToTheScript replied:
I had that in there but must've accidentally written over my previous save. Adding it back in.

franher007 commented on Reflection/NPC.simba:
```diff
@@ -201,3 +222,234 @@ begin
+
+   writeln(R_NPCGetMSCoordsClosest('Chicken'));
+*)
+Function R_NPCGetMSCoordsClosest(Name: String): TPoint;
```
Similar suggestion here

StickToTheScript committed:
Added minor changes

Brandon-T commented:
For a lot of these NPC functions, it might be better to just make a:

```pascal
type TNPC = record
    name: String;
    tile: RSTile;
    combat_level: Int32;
    etc..
end;
```

and just have a function like `GetNPCs(ID: Int32): Array of TNPC` and `GetAllNPCs: Array of TNPC`

That way, you don't have so much repeated code? Example: https://github.com/SRL/SRL-Reflection-2/blob/master/lib/animable/characters/NPCs.simba#L113

StickToTheScript committed:
Minor fixes.

StickToTheScript committed:
Updated NPC.simba and Mainscreen.simba

Added a record to simplify the NPC functions.

Added resizable option for getting bounds of mainscreen.

StickToTheScript commented:
> For a lot of these NPC functions, it might be better to just make a:
> 
> ```pascal
> type TNPC = record
>     name: String;
>     tile: RSTile;
>     combat_level: Int32;
>     etc..
> end;
> ```
> 
> and just have a function like `GetNPCs(ID: Int32): Array of TNPC` and `GetAllNPCs: Array of TNPC`
> 
> That way, you don't have so much repeated code? Example: https://github.com/SRL/SRL-Reflection-2/blob/master/lib/animable/characters/NPCs.simba#L113

Thoughts on something like the new update?

I left in the isVisible functions and reworked the R_NPCGetMSCoordsSort.

StickToTheScript committed:
Added TRSNPC and TRSObjects

Changed TNPC to TRSNPC

Added TRSObjects.

StickToTheScript committed:
Reordered Reflection.simba

StickToTheScript committed:
Added AnimationID to TRSNPC.

StickToTheScript committed:
Added Model to TRSObject

StickToTheScript committed:
Added Health to TRSNPC

StickToTheScript committed:
Added LocalTile to TRSNPC

StickToTheScript committed:
Added LocalTile ot TRSObject

StickToTheScript committed:
Update Objects.simba
